Sources/AblyChat/DefaultMessages.swift

```diff
@@ -0,0 +1,237 @@
+    public nonisolated func subscribeToDiscontinuities() -> Subscription<ARTErrorInfo> {
+        fatalError("not implemented")
+    }
+
+    public nonisolated func discontinuityDetected(reason: ARTErrorInfo?) {
+        print("Discontinuity detected: \(reason ?? .createUnknownError())")
+    }
+
+    private func getBeforeSubscriptionStart(_ uuid: UUID, params: QueryOptions) async throws -> any PaginatedResult<Message> {
+        guard let subscriptionPoint = subscriptionPoints[uuid]?.timeserial else {
+            throw ARTErrorInfo.create(
+                withCode: 40000,
+                status: 400,
+                message: "cannot query history; listener has not been subscribed yet"
+            )
+        }
+
+        // (CHA-M5j) If the end parameter is specified and is more recent than the subscription point timeserial, the method must throw an ErrorInfo with code 40000.
+        let parseSerial = try? DefaultTimeserial.calculateTimeserial(from: subscriptionPoint)
+        if let end = params.end, end > parseSerial?.timestamp ?? 0 {
+            throw ARTErrorInfo.create(
+                withCode: 40000,
+                status: 400,
+                message: "cannot query history; end time is after the subscription point of the listener"
+            )
+        }
+
+        // (CHA-M5f) This method must accept any of the standard history query options, except for direction, which must always be backwards.
+        var queryOptions = params
+        queryOptions.orderBy = .newestFirst // newestFirst is equivalent to backwards
+
+        // (CHA-M5g) The subscribers subscription point must be additionally specified (internally, by us) in the fromSerial query parameter.
+        queryOptions.fromSerial = subscriptionPoint
+
+        return try await chatAPI.getMessages(roomId: roomID, params: queryOptions)
+    }
+
+    private func handleChannelEvents(roomId _: String) {
+        // (CHA-M5c) If a channel leaves the ATTACHED state and then re-enters ATTACHED with resumed=false, then it must be assumed that messages have been missed. The subscription point of any subscribers must be reset to the attachSerial.
+        channel.on(.attached) { [weak self] stateChange in
+            Task {
+                do {
+                    try await self?.handleAttach(fromResume: stateChange.resumed)
+                    return
+                } catch {
+                    throw ARTErrorInfo.create(from: error)
+                }
+            }
+        }
+
+        // (CHA-M4d) If a channel UPDATE event is received and resumed=false, then it must be assumed that messages have been missed. The subscription point of any subscribers must be reset to the attachSerial.
+        channel.on(.update) { [weak self] stateChange in
+            if stateChange.current == .attached, stateChange.previous == .attached {
```

Comment:
that doesn't make sense semantically to me... surely something can update from 1 value to another? 

this is also the same as chat-js

Reply:
> surely something can update from 1 value to another?

You could argue the naming is confusing, but that's not what the UPDATE event means. See https://ably.com/docs/channels?lang=javascript#listen-for-state:

> The Channel object can also emit one event that is not a state change: an update event. This happens when there’s a change to channel conditions for which the channel state doesn’t change.

When the channel _changes_ state, you get an event whose name is the new state, not an `UPDATE`.

> this is also the same as chat-js

We should be following what the spec says, not what chat-js does, otherwise there's not much point in having a spec. If the spec is wrong or insufficent, that needs to be addressed.

Reply:
ahh interesting, the name is definitely confusing! I've gotten rid of the redundant check